pkg/hooks/manager.go

```diff
@@ -0,0 +1,228 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package hooks
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	"kubevirt.io/kubevirt/pkg/log"
+	domainSchema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	virtwrapApi "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+type callackClient struct {
+	SocketPath          string
+	Version             string
+	subsribedHookPoints []*hooksInfo.HookPoint
+}
+
+var manager *Manager
+var once sync.Once
+
+type Manager struct {
+	callbacksPerHookPoint map[string][]*callackClient
+}
+
+func GetManager() *Manager {
+	once.Do(func() {
+		manager = &Manager{callbacksPerHookPoint: make(map[string][]*callackClient)}
+	})
+	return manager
+}
+
+func (m *Manager) Collect(numberOfRequestedHookSidecars uint, timeout time.Duration) error {
+	callbacksPerHookPoint, err := collectSideCarSockets(numberOfRequestedHookSidecars, timeout)
+	if err != nil {
+		return err
+	}
+	log.Log.Info("Collected all requested hook sidecar sockets")
+
+	sortCallbacksPerHookPoint(callbacksPerHookPoint)
+	log.Log.Infof("Sorted all collected sidecar sockets per hook point based on their priority and name: %v", callbacksPerHookPoint)
+
+	m.callbacksPerHookPoint = callbacksPerHookPoint
+
+	return nil
+}
+
+// TODO: Handle sockets in parallel, when a socket appears, run a goroutine trying to read Info from it
+func collectSideCarSockets(numberOfRequestedHookSidecars uint, timeout time.Duration) (map[string][]*callackClient, error) {
+	callbacksPerHookPoint := make(map[string][]*callackClient)
+	processedSockets := make(map[string]bool)
+
+	timeoutCh := time.After(timeout)
```

Comment:
Minor point, but wouldn't you want to give each side car it's own timeout. This says for *all* sidecars, wait this time. 

To me, it seems like the timeout should be per side car initialization.

Reply:
ah nvm, i think this is just to collect the dockets,

Reply:
That sounds reasonable. I'd like to make the Collect function to run in parallel, but for the meantime this sounds like a good solution. What do you think @davidvossel ?

Reply:
Nevermind, actually I think that the timeout should be shared. All sidecars starts in parallel, so we are waiting for there images to boot. The Info() call should take no time.

Reply:
yeah, this is just the global collection timeout. The timeout is pretty conservative (5 minutes I think). It's just to ensure we eventually fail if a sidecar doesn't come online successfully.